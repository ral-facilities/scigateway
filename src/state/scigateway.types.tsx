import { Theme } from '@mui/material';
import { Step } from 'react-joyride';
import { LogoState } from './state.types';

export const AuthSuccessType = 'scigateway:auth_success';
export const AuthFailureType = 'scigateway:auth_failure';
export const AutoLoginSuccessType = 'scigateway:autologin_success';
export const LoadingAuthType = 'scigateway:loading_auth';
export const LoadedAuthType = 'scigateway:loaded_auth';
export const LoadAuthProviderType = 'scigateway:loading_auth_provider';
export const LoadScheduledMaintenanceStateType =
  'scigateway:load_scheduled_maintenance_state';
export const LoadMaintenanceStateType = 'scigateway:load_maintenance_state';
export const InvalidateTokenType = 'scigateway:api:invalidate_token';
export const ConfigureStringsType = 'scigateway:configure_strings';
export const LoginType = 'scigateway:login';
export const NotificationType = 'scigateway:api:notification';
export const RegisterRouteType = 'scigateway:api:register_route';
export const RequestPluginRerenderType = 'scigateway:api:plugin_rerender';
export const SendThemeOptionsType = 'scigateway:api:send_themeoptions';
export const LoadDarkModePreferenceType =
  'scigateway:load_dark_mode_preference';
export const LoadHighContrastModePreferenceType =
  'scigateway:load_high_contrast_mode_preference';
export const SignOutType = 'scigateway:signout';
export const ResetAuthStateType = 'scigateway:reset_auth_state';
export const BroadcastSignOutType = 'scigateway:api:signout';
export const ToggleDrawerType = 'scigateway:toggledrawer';
export const DismissNotificationType = 'scigateway:dismissnotification';
export const ConfigureFeatureSwitchesType = 'scigateway:feature_switches';
export const AddNotificationsType = 'scigateway:add_notification';
export const SiteLoadingType = 'scigateway:site_loading';
export const ConfigureAnalyticsType = 'scigateway:configure_analytics';
export const InitialiseAnalyticsType = 'scigateway:initialise_analytics';
export const ToggleHelpType = 'scigateway:toggle_help';
export const AddHelpTourStepsType = 'scigateway:add_help_tour_steps';
export const RegisterHomepageUrlType = 'scigateway:register_homepage_url';
export const CustomLogoType = 'scigateway:custom_logo';
export const CustomNavigationDrawerLogoType =
  'scigateway:custom_navigation_drawer_logo';
export const CustomAdminPageDefaultTabType =
  'scigateway:custom_admin_default_tab';
export const RegisterContactUsAccessibilityFormUrlType =
  'scigateway:contact_us_accessibility_form_url';
export const CustomPrimaryColourType = 'scigateway:custom_primary_colour';

export const scigatewayRoutes = {
  home: '/',
  contact: '/contact',
  help: '/help',
  admin: '/admin',
  login: '/login',
  logout: '/logout',
  cookies: '/cookies',
  accessibility: '/accessibility',
};
export const adminRoutes = {
  maintenance: '/admin/maintenance',
  download: '/admin/download',
};

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export type AppStrings = Record<string, string>;

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export type ApplicationStrings = Record<string, AppStrings>;

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showHelpPageButton: boolean;
  singlePluginLogo: boolean;
}

export interface HomepageUrlPayload {
  homepageUrl: string;
}

export interface CustomLogoPayload {
  logo: string;
}

export interface CustomNavigationDrawerLogoPayload {
  navigationDrawerLogo: LogoState;
}

export interface CustomAdminPageDefaultTabPayload {
  adminPageDefaultTab: 'maintenance' | 'download';
}

export interface CustomPrimaryColourPayload {
  primaryColour: string;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  unauthorised?: boolean;
  hideFromMenu?: boolean;
  admin?: boolean;
  order: number;
  helpText?: string;
  logoLightMode?: string;
  logoDarkMode?: string;
  logoAltText?: string;
  helpSteps?: { target: string; content: string }[];
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  unauthorised?: boolean;
  hideFromMenu?: boolean;
  admin?: boolean;
  order: number;
  helpText?: string;
  logoLightMode?: string;
  logoDarkMode?: string;
  logoAltText?: string;
  helpSteps?: { target: string; content: string }[];
}

export type GroupedPlugins = Record<string, PluginConfig[]>;

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface SendThemeOptionsPayload {
  theme: Theme;
  broadcast: boolean;
}

export interface LoadDarkModePreferencePayload {
  darkMode: boolean;
}

export interface LoadHighContrastModePreferencePayload {
  highContrastMode: boolean;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string | null;
  authUrl?: string;
  autoLogin?: boolean;
}

export interface ScheduledMaintenanceStatePayLoad {
  scheduledMaintenance: ScheduledMaintenanceState;
}

export interface ScheduledMaintenanceState {
  show: boolean;
  message: string;
  severity?: 'success' | 'warning' | 'error' | 'information';
}

export interface MaintenanceStatePayLoad {
  maintenance: MaintenanceState;
}

export interface MaintenanceState {
  show: boolean;
  message: string;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}

export interface ContactUsAccessibilityFormUrlPayload {
  contactUsAccessibilityFormUrl: string;
}
