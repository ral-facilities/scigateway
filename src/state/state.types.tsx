import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';
import { Step } from 'react-joyride';
import {
  ApplicationStrings,
  PluginConfig,
  FeatureSwitches,
  ScheduledMaintenanceState,
  MaintenanceState,
} from './scigateway.types';
import { RouterState } from 'connected-react-router';

export interface Plugin {
  name: string;
  src: string;
  enable: boolean;
  unauthorised?: boolean;
  location: 'main' | 'left' | 'right';
}

export interface ScigatewayNotification {
  message: string;
  severity: string;
}

export interface ScigatewayState {
  notifications: ScigatewayNotification[];
  plugins: PluginConfig[];
  drawerOpen: boolean;
  siteLoading: boolean;
  showHelp: boolean;
  helpSteps: Step[];
  authorisation: AuthState;
  res?: ApplicationStrings;
  features: FeatureSwitches;
  analytics?: AnalyticsState;
  darkMode: boolean;
  highContrastMode: boolean;
  homepageUrl?: string;
  logo?: string;
  scheduledMaintenance: ScheduledMaintenanceState;
  maintenance: MaintenanceState;
  navigationDrawerLogo?: LogoState;
  adminPageDefaultTab?: 'maintenance' | 'download';
  contactUsAccessibilityFormUrl?: string;
  primaryColour?: string;
}

export interface StateType {
  scigateway: ScigatewayState;
  router: RouterState;
}

export interface ActionType<T> {
  type: string;
  payload: T;
}

export type ThunkResult<R> = ThunkAction<R, StateType, null, AnyAction>;

export interface User {
  username: string;
  isAdmin: boolean;
  avatarUrl: string;
}

export interface ICATAuthenticator {
  mnemonic: string;
  keys: { name: string; hide?: boolean }[];
  friendly?: string;
  admin?: boolean;
}

export interface AuthProvider {
  isLoggedIn: () => boolean;
  isAdmin: () => boolean;
  logOut: () => void;
  logIn: (username: string, password: string) => Promise<void>;
  verifyLogIn: () => Promise<void>;
  refresh: () => Promise<void>;
  fetchScheduledMaintenanceState?: () => Promise<ScheduledMaintenanceState>;
  setScheduledMaintenanceState?: (
    scheduledMaintenanceState: ScheduledMaintenanceState
  ) => Promise<string | void>;
  fetchMaintenanceState?: () => Promise<MaintenanceState>;
  setMaintenanceState?: (
    maintenanceState: MaintenanceState
  ) => Promise<string | void>;
  redirectUrl: string | null;
  authUrl: string | undefined;
  user: User | null;
  mnemonic?: string;
  autoLogin?: () => Promise<void>;
}

export interface AuthState {
  failedToLogin: boolean;
  signedOutDueToTokenInvalidation: boolean;
  loading: boolean;
  provider: AuthProvider;
}

export interface AnalyticsState {
  id: string;
  initialised: boolean;
}

export interface LogoState {
  light: string;
  dark: string;
  altTxt: string;
}
