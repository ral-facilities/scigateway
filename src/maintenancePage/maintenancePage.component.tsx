import React from 'react';
import Typography from '@mui/material/Typography';
import { connect } from 'react-redux';
import { StateType } from '../state/state.types';
import { Box } from '@mui/material';

interface MaintenancePageProps {
  message: string;
}

const MaintenancePage = (props: MaintenancePageProps): React.ReactElement => {
  return (
    <Box
      sx={{
        padding: 2,
        backgroundColor: 'background.default',
        textAlign: 'center',
      }}
    >
      <Typography
        variant="h3"
        sx={{ fontWeight: 'bold', color: 'secondary.main' }}
      >
        Maintenance
      </Typography>
      <Box
        sx={{
          display: 'flex',
          flexDirection: 'column',
          marginTop: 2,
          marginBottom: 2,
          color: 'text.primary',
        }}
      >
        <Typography variant="h4">{props.message}</Typography>
      </Box>
    </Box>
  );
};

const mapStateToProps = (state: StateType): MaintenancePageProps => ({
  message: state.scigateway.maintenance.message,
});

export default connect(mapStateToProps)(MaintenancePage);
