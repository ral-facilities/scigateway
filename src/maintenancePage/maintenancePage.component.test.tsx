import React from 'react';
import configureStore from 'redux-mock-store';
import { default as MaintenancePage } from './maintenancePage.component';
import { StateType } from '../state/state.types';
import { authState, initialState } from '../state/reducers/scigateway.reducer';
import { render } from '@testing-library/react';

describe('Maintenance page component', () => {
  let mockStore;
  let state: StateType;

  beforeEach(() => {
    mockStore = configureStore();
    state = {
      scigateway: { ...initialState, authorisation: { ...authState } },
    };
    state.scigateway.maintenance.message = 'test';
  });

  it('should render correctly', () => {
    const testStore = mockStore(state);
    const { asFragment } = render(<MaintenancePage store={testStore} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
