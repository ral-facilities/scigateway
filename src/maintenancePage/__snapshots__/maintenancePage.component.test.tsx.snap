// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Maintenance page component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-hh6rrp"
  >
    <h3
      class="MuiTypography-root MuiTypography-h3 css-11q8ert-MuiTypography-root"
    >
      Maintenance
    </h3>
    <div
      class="MuiBox-root css-9c2a3a"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
      >
        test
      </h4>
    </div>
  </div>
</DocumentFragment>
`;
